AssertionError: assertion failed: leak

## Compiler version

- 3.3.1-RC1-bin-20230216-2507577-NIGHTLY
- 3.3.0-RC3
- 3.2.2

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
package example

import scala.quoted.Expr
import scala.quoted.Quotes
import scala.quoted.Type

object Main {

  inline def f[T]: Int = ${ g[T] }

  def g[T: Type](using Quotes): Expr[Int] = {
    val a = {
      Type.of[T] match {
        case '[Option[x]] =>
          classOf[x]
      }
      3
    }

    Expr(7)
  }
}
```

## Output (click arrow to expand)

<details>

```scala
[error] java.lang.AssertionError: assertion failed: leak: x in {
[error]   classOf[x]:(classOf[x] : Class[x])
[error] }
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.typer.Typer.ensureNoLocalRefs(Typer.scala:1162)
[error] dotty.tools.dotc.typer.Typer.caseRest$1(Typer.scala:1778)
[error] dotty.tools.dotc.typer.Typer.typedCase(Typer.scala:1794)
[error] dotty.tools.dotc.typer.Typer.typedCases$$anonfun$1(Typer.scala:1724)
[error] dotty.tools.dotc.core.Decorators$.loop$1(Decorators.scala:94)
[error] dotty.tools.dotc.core.Decorators$.mapconserve(Decorators.scala:110)
[error] dotty.tools.dotc.typer.Typer.typedCases(Typer.scala:1726)
[error] dotty.tools.dotc.typer.Typer.$anonfun$26(Typer.scala:1716)
[error] dotty.tools.dotc.typer.Applications.harmonic(Applications.scala:2319)
[error] dotty.tools.dotc.typer.Applications.harmonic$(Applications.scala:352)
[error] dotty.tools.dotc.typer.Typer.harmonic(Typer.scala:115)
[error] dotty.tools.dotc.typer.Typer.typedMatchFinish(Typer.scala:1716)
[error] dotty.tools.dotc.typer.Typer.typedMatch(Typer.scala:1650)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2942)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2991)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3111)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3130)
[error] dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1096)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1100)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2936)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2991)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3174)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1619)
[error] dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1609)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1619)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1872)
[error] dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:249)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1(Namer.scala:1872)
[error] dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1880)
[error] dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1898)
[error] dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1899)
[error] dotty.tools.dotc.typer.Namer.inferredResultType(Namer.scala:1910)
[error] dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1657)
[error] dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1664)
[error] dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:783)
[error] dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:922)
[error] dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:810)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:174)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:187)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:189)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:390)
[error] dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2869)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2894)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2990)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3084)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3130)
[error] dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1096)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1100)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2936)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2991)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3174)
[error] dotty.tools.dotc.typer.Typer.$anonfun$49(Typer.scala:2382)
[error] dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:249)
[error] dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2382)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2904)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2990)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3084)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3130)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2562)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2920)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2990)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3084)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3130)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2692)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2961)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2991)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3058)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:3062)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3174)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:44)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$adapted$1(TyperPhase.scala:54)
[error] scala.Function0.apply$mcV$sp(Function0.scala:42)
[error] dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:437)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:54)
[error] dotty.tools.dotc.typer.TyperPhase.runOn$$anonfun$3(TyperPhase.scala:88)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:88)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:247)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:263)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:271)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:280)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:280)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:195)
[error] dotty.tools.dotc.Run.compile(Run.scala:179)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:179)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:177)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:463)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:418)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:405)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:171)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2363)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2313)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:30)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:30)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2311)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:829)
```
</details>

#### Minimization

```scala
def test =
  type y
  classOf[y]
```

<details>
<summary>Failure</summary>

```scala
3 |  classOf[y]
  |          ^
  |          y is not a class type
java.lang.AssertionError: assertion failed: leak: y in {
  type y >: Nothing <: Any
  classOf[y]:(classOf[y] : Class[y])
} while typechecking tests/pos-macros/i16954.scala
exception occurred while typechecking tests/pos-macros/i16954.scala
exception occurred while compiling tests/pos-macros/i16954.scala
java.lang.AssertionError: assertion failed: leak: y in {
  type y >: Nothing <: Any
  classOf[y]:(classOf[y] : Class[y])
} while compiling tests/pos-macros/i16954.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: leak: y in {
  type y >: Nothing <: Any
  classOf[y]:(classOf[y] : Class[y])
}
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.typer.Typer.ensureNoLocalRefs(Typer.scala:1165)
        at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1121)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2955)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3010)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3077)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3081)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3193)
        at dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1645)
        at dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1635)
        at dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1645)
        at dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1898)
        at dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:249)
        at dotty.tools.dotc.typer.Namer.typedAheadRhs$1(Namer.scala:1898)
        at dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1906)
        at dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1924)
        at dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1925)
        at dotty.tools.dotc.typer.Namer.inferredResultType(Namer.scala:1936)
        at dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1683)
        at dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1690)
        at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1781)
        at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:791)
        at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:926)
        at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:814)
        at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:174)
        at dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:187)
        at dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:189)
        at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:390)
        at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2888)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2913)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3009)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3077)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3081)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3103)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3149)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2581)
        at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2935)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2939)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3009)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3077)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3081)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3103)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3149)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2711)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2980)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3010)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3077)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3081)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3193)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:44)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$adapted$1(TyperPhase.scala:54)
        at scala.Function0.apply$mcV$sp(Function0.scala:42)
        at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:437)
        at dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:54)
        at dotty.tools.dotc.typer.TyperPhase.runOn$$anonfun$3(TyperPhase.scala:88)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:88)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:247)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:263)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:271)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:280)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:280)
        at dotty.tools.dotc.Run.compileSources(Run.scala:195)
        at dotty.tools.dotc.Run.compile(Run.scala:179)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
        at dotty.tools.dotc.Driver.process(Driver.scala:195)
        at dotty.tools.dotc.Driver.process(Driver.scala:163)
        at dotty.tools.dotc.Driver.process(Driver.scala:175)
        at dotty.tools.dotc.Driver.main(Driver.scala:205)
        at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -5736,6 +5736,12 @@ object Types {
         case tp @ SuperType(thistp, supertp) =>
           derivedSuperType(tp, this(thistp), this(supertp))
 
+        case tp @ ConstantType(const @ Constant(_: Type)) =>
+          val classType = const.tpe
+          val classType1 = this(classType)
+          if classType eq classType1 then tp
+          else classType1
+
         case tp: LazyRef =>
           LazyRef { refCtx =>
             given Context = refCtx

--- a/compiler/src/dotty/tools/dotc/printing/Formatting.scala
+++ b/compiler/src/dotty/tools/dotc/printing/Formatting.scala
@@ -71,6 +71,16 @@ object Formatting {
       given Show[TypeComparer.ApproxState] with
         def show(x: TypeComparer.ApproxState) = TypeComparer.ApproxState.Repr.show(x)
 
+      given Show[ast.TreeInfo.PurityLevel] with
+        def show(x: ast.TreeInfo.PurityLevel) = x match
+          case ast.TreeInfo.Path           => "PurityLevel.Path"
+          case ast.TreeInfo.Pure           => "PurityLevel.Pure"
+          case ast.TreeInfo.Idempotent     => "PurityLevel.Idempotent"
+          case ast.TreeInfo.Impure         => "PurityLevel.Impure"
+          case ast.TreeInfo.PurePath       => "PurityLevel.PurePath"
+          case ast.TreeInfo.IdempotentPath => "PurityLevel.IdempotentPath"
+          case _                           => s"PurityLevel(${x.x})"
+
       given Show[Showable]                            = ShowAny
       given Show[Shown]                               = ShowAny
       given Show[Int]                                 = ShowAny

--- a/tests/pos/i16954.scala
+++ b/tests/pos/i16954.scala
@@ -1,0 +1,18 @@
+class Test:
+  def test =
+    classOf[Test]
+
+  def blck =
+    class Blck
+    val cls = classOf[Blck]
+    cls
+
+  def expr =
+    class Expr
+    classOf[Expr] // was: "assertion failed: leak: Expr in { [..] }" crash
+
+object Test extends Test:
+  def main(args: Array[String]): Unit =
+    assert(test.getName == "Test",        test.getName)
+    assert(blck.getName == "Test$Blck$1", blck.getName)
+    assert(expr.getName == "Test$Expr$1", expr.getName)

